subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DefaultPendingDependenciesHandler.java

```diff
@@ -31,9 +31,10 @@
  * optional dependencies, then the optional configuration becomes effectively empty. To avoid this, we ignore the state of optional dependencies if they belong to the "optional" configuration.
  */
 public class DefaultPendingDependenciesHandler implements PendingDependenciesHandler {
-    private final PendingDependenciesState pendingDependencies = new PendingDependenciesState();
     private final DependencySubstitutionApplicator dependencySubstitutionApplicator;
     private final ComponentSelectorConverter componentSelectorConverter;
+    // Create on demand
+    private PendingDependenciesState pendingDependencyTracker = new PendingDependenciesState();
```

Comment:
:x: I think this `new PendingDependenciesState();` sneaked in here and should not be here ;)

Reply:
Yeah, that might explain why the performance tests are still failing! I removed the `final` and added a comment: just didn't actually remove the initializer!

Reply:
This is now fixed: I amended the original commit and force pushed.

Reply:
No, that doesn't work at all. It seems like our pending dependencies support relies on us tracking all of the non-pending dependencies that are seen before we detect any pending dependencies.

Reply:
Damn... I didn't think of that. Sorry. 😞